tests/io/vasp/test_inputs.py

```diff
@@ -1035,51 +1033,58 @@ def test_functional_types(self):
 
         assert self.psingle.symbol == "Mn_pv"
 
-    def test_identify_potcar(self):
-        filename = f"{TEST_FILES_DIR}/POT_GGA_PAW_PBE_54/POTCAR.Fe.gz"
+    def test_is_valid(self):
+        Fe_psp_path = f"{TEST_FILES_DIR}/POT_GGA_PAW_PBE_54/POTCAR.Fe.gz"
 
-        psingle = PotcarSingle.from_file(filename)
-        assert "PBE_54" in psingle.identify_potcar()[0]
-        assert "Fe" in psingle.identify_potcar()[1]
+        psingle = PotcarSingle.from_file(Fe_psp_path)
+        assert psingle.is_valid
+
+        # corrupt the file
+        assert psingle.keywords["RCORE"] == 2.3
+        psingle.keywords["RCORE"] = 2.2
+        assert not psingle.is_valid
 
-    def test_potcar_hash_warning(self):
+        psingle.keywords.pop("RCORE")
+        assert not psingle.is_valid
+
+        psingle.data = psingle.data.replace("RCORE  =    2.300", "RCORE  =    2.200")
+        assert not psingle.is_valid
+
+    def test_unknown_potcar_warning(self):
```

Comment:
I think rather than storing a new POTCAR file with just 1 changed line, it should be easy to invalidate a `PotcarSingle` instance on the fly. But actually this turned out to be pretty hard. You might say this is a contrived example since people shouldn't change `PotcarSingle` objects once they've been read from a file. But fwiw, I can currently reassign

```
psingle.keywords["RCORE"] = 2.2
```

remove it altogether

```
psingle.keywords.pop("RCORE")
```

and modify the underlying string data

```
psingle.data = psingle.data.replace("RCORE  =    2.300", "RCORE  =    2.200")
```

all without invalidating the `PotcarSingle` instance. See the new `assert`s I added. I think it would make sense for all of them to pass but they currently fail.

Reply:
This is because the header info is taken from `self.PSCTR` and not the `parse_functions` attributes of PotcarSingle. For each such attribute like RCORE (the ones in `self.parse_functions`), the key/value are stored both in `self.keywords` and `self.PSCTR`

So if you do instead `self.PSCTR['RCORE'] = 2.2`, the second `psingle.is_valid` will fail. Or, if you did 
`psingle.data = psingle.data.replace('local part','local port')`
the second `psingle.is_valid` will also fail.

Hesitant to change how this is done (storing the same data in three different attributes of PotcarSingle) because it might break something that relies on it

Reply:
> Hesitant to change how this is done (storing the same data in three different attributes of PotcarSingle) because it might break something that relies on it

I don't think we need to be hesitant here. having redundant data in `PotcarSingle.keywords` and `PotcarSingle.PSCTR` is needlessly complicated, goes against the SSOT coding principle and now would be a great time to clean this up imo.
@shyuep @mkhorton Any concerns about breaking reworking `PotcarSingle` to e.g. remove `PSCTR` and turn `keywords` into a `@property` so they're computed on the fly and always in sync with the underlying `data` string?

Reply:
I agree, having redundant data doesn't make sense for the reasons you state. My only hesitation here is that I wouldn't want to impose scope creep on this PR for @esoteric-ephemera, since it's probably not strictly necessary to change it for this PR to achieve its goal of better validation. While the `PotcarSingle` should be immutable, I think the risk of someone mutating it in practice are quite low.

More broadly, I've never had a good understanding of the internals of the POTCAR file. If anyone did want to spend more time on the `PotcarSingle` class, I'd love to see more documented `@property` etc. for some of these keywords so that people can better understand what the POTCAR files they're using are actually specifying.

Reply:
I merged `PotcarSingle.keywords` and `PotcarSingle.PSCTR`, now just called `PotcarSingle.keywords` since `PSCTR` has no obvious meaning. Maybe we should rename `PotcarSingle.keywords` to `PotcarSingle.header_keywords` for added clarity?